flink-connectors/flink-connector-kafka-0.9/src/test/java/org/apache/flink/streaming/connectors/kafka/internal/Flink9349Test.java

```diff
@@ -0,0 +1,207 @@
+		final MultiShotLatch blockerLatch = new MultiShotLatch();
+
+		KafkaConsumer<?, ?> mockConsumer = mock(KafkaConsumer.class);
+		when(mockConsumer.poll(anyLong())).thenAnswer(new Answer<ConsumerRecords<?, ?>>() {
+
+			@Override
+			public ConsumerRecords<?, ?> answer(InvocationOnMock invocation) throws InterruptedException {
+				sync.trigger();
+				blockerLatch.await();
+				return ConsumerRecords.empty();
+			}
+		});
+
+		doAnswer(new Answer<Void>() {
+			@Override
+			public Void answer(InvocationOnMock invocation) {
+				blockerLatch.trigger();
+				return null;
+			}
+		}).when(mockConsumer).wakeup();
+
+		// make sure the fetcher creates the mock consumer
+		whenNew(KafkaConsumer.class).withAnyArguments().thenReturn(mockConsumer);
+
+		// ----- create the test fetcher -----
+
+		@SuppressWarnings("unchecked")
+		SourceFunction.SourceContext<String> sourceContext = mock(SourceFunction.SourceContext.class);
+		Map<KafkaTopicPartition, Long> partitionsWithInitialOffsets =
+			Collections.singletonMap(new KafkaTopicPartition("test", 42), KafkaTopicPartitionStateSentinel.GROUP_OFFSET);
+		KeyedDeserializationSchema<String> schema = new KeyedDeserializationSchemaWrapper<>(new SimpleStringSchema());
+
+		final Kafka09Fetcher<String> fetcher = new Kafka09Fetcher<>(
+			sourceContext,
+			partitionsWithInitialOffsets,
+			null, /* periodic watermark extractor */
+			null, /* punctuated watermark extractor */
+			new TestProcessingTimeService(),
+			10, /* watermark interval */
+			this.getClass().getClassLoader(),
+			"task_name",
+			schema,
+			new Properties(),
+			0L,
+			new UnregisteredMetricsGroup(),
+			new UnregisteredMetricsGroup(),
+			false);
+
+		// ----- run the fetcher -----
+
+		final AtomicReference<Throwable> error = new AtomicReference<>();
+		int fetchTasks = 2;
+		final CountDownLatch latch = new CountDownLatch(fetchTasks);
+		ExecutorService service = Executors.newFixedThreadPool(fetchTasks + 1);
+
+		service.submit(new Thread("fetcher runner ") {
+			@Override
+			public void run() {
+				try {
+					latch.await();
+					fetcher.runFetchLoop();
+				} catch (Throwable t) {
+					error.set(t);
+				}
+			}
+		});
+		for (int i = 0; i < fetchTasks; i++) {
+			service.submit(new Thread("add partitions " + i) {
+
+				@Override
+				public void run() {
+					try {
+						List<KafkaTopicPartition> newPartitions = new ArrayList<>();
+						for (int i = 0; i < 1000; i++) {
+							newPartitions.add(testPartition);
+						}
+						fetcher.addDiscoveredPartitions(newPartitions);
+						latch.countDown();
+						//latch.await();
+						for (int i = 0; i < 100; i++) {
+							fetcher.addDiscoveredPartitions(newPartitions);
+							Thread.sleep(1L);
+						}
+					} catch (Throwable t) {
+						error.set(t);
+					}
+				}
+			});
+		}
+
+		service.awaitTermination(1L, TimeUnit.SECONDS);
+
+		// wait until the fetcher has reached the method of interest
+		sync.await();
+
+		// ----- trigger the offset commit -----
```

Comment:
I don't think this is required for the scope of interest of this test, is it?

Reply:
thank you 
you are right I removed it and some other useless stuff from this test